### ISSUE TITLE
Help with multiple search and some regex

### ISSUE THREAD

**Issue by Jorman** [Jackett/Jackett#3498]
Hi, I just noticed that in most of "italian" tracker there's a problem.

Like you know I don't know very well the yaml language so I can only try and search but this time I think some help is needed to fix forever this bad situation.

So let's go in order

1. On most tracker I find mixed release, I know that is not Jackett fault, but I can't contact every release to try to tell him how to make a release :). The problem is that: there's not a standard, some times they put s01e01 some times 1x01, even during a season, I find some episode in one format and others in the other one... A mess!!!!!
So, is possible to make 2 consecutive search with Jackett? My idea is to make one search with
S01 or S01E01 and another search with
1x01 or 1 or "sometexthere" 1
All of this, with appropriate keywordsfilters and maybe only when a tv show are searched!

2. I think I've some problem with keywordsfilters, but first a simple question, I've to set the keywordsfilters before the `search: "{{ .Keywords }}"` or after? On the definition format into the wiki section I saw it after, but almost all script that use keywordsfilters specify that before.
See this code
```
  search:
    paths:
      - path: /index.php
    keywordsfilters:
      - name: re_replace
        args: ["\\bS(\d+)\\b", "$1"] # remove season tag without episode (search doesn't support it)
    inputs:
      search: "{{ .Keywords }}"
      category: "{{range .Categories}}{{.}};{{end}}"
      page: "torrents"
      active: 0
```
In theory I transform a search like
royasl s4 into royasl 4
The only problem is that I've to remove the leading 0 because S01 must be 1 not 01, any idea?
UPDATE -> maybe this handle better the leading 0
https://regex101.com/r/1jOBem/2/
Anyway this works in theory, see this
https://regex101.com/r/1jOBem/1/
But if I try it into Jackett, no results!
![1](https://user-images.githubusercontent.com/5203988/43485164-be7a1c34-9510-11e8-9769-73633dc2b6ee.PNG)
![2](https://user-images.githubusercontent.com/5203988/43485173-c3a77e2c-9510-11e8-9fb2-ce88d8bb90f6.PNG)

What I miss?

**Comment by kaso17** [comment]
the /locationorder of the keywordsfilters doesn't matter.
in your example `\d+` isn't double escaped.
to remove leading zeroes `\bS0*(\d+)\b` should work.
so it should be `["\\bS0*(\\d+)\\b", "$1"]`

**Comment by Jorman** [comment]
Yep, noticed the double escape! Already fixed, but seems not working. I search
royals s04 but no results
if I search royals 4 I got 13 results

What I miss?

What about the double search to handle the people who make the wrong release name? Is possible?

**Comment by kaso17** [comment]
The reges is workign fine for me.
Please check the enhanced log to see if the correct request is sent.

**Comment by kaso17** [comment]
Regarding the multiple searches: no it's not possible with cardigann definitions

**Comment by Jorman** [comment]
So if I put 2 time the search cardigan don't aggregate the results, right?
Any idea to resolve the bad releasers?

About the regex here my log, I hope can help to understand why there's this problem
https://pastebin.com/Tc4k17q3
and here the settings into the definition file
https://pastebin.com/H6aEW1t8

As you can see I only use the keywordsfilters filter before the search.

J

**Comment by kaso17** [comment]
Looks fine to me. I don't see any obvious problem. Do you have an invite for girotorrent so I can test it myself or which other trackers are affected.

Which second search request do you intend to send?
If stripping S/E/leading zeros is enough to return all relevant results that should be fine.
Then you just have to normalize the various release patterns via filters (title field) so they all become the standard SXXEYY format (otherwise sonarr, etc. won't pick them up).
Additionally a andmatch filter might be useful.

In case of other indexers which only support searching on word boundaries the simples solution for now is to  completely remove the season/episode form the keywords, normalize the results and and filter them  using andmatch. Specifying multiple identical paths with different search keywords is a more complex workaround. But in this case you'll be limited to a single re_replace operation per request (based on the previous keywordsfilters).

**Comment by Jorman** [comment]
I really don't know the andmatch filter, I've to see how it works and how can be useful in this scenario

So, we've 2 discussion on the same tread:
1. Why you said that looks fine? I don't have an invite yet, but is the same on Il Corsaro Blu that is not private. On il Corsaro Blu for example if I search Grimm s06 I obtain 3 results
![1](https://user-images.githubusercontent.com/5203988/43609219-adef5548-96a3-11e8-8508-6c60353c0ef0.PNG)
But my expectation is more than 3
![2](https://user-images.githubusercontent.com/5203988/43609251-c82e3ece-96a3-11e8-9f1a-2cb3c9ec2004.PNG)
And here the code that tell to remove S0 before do the search, but seems not working
```
  search:
    paths:
      - path: /index.php
    keywordsfilters:
      - name: re_replace
        args: ["\\bS0*(\\d+)\\b", "$1"] # remove season tag without episode (search doesn't support it)
        # args: ["S[0-9]{2}([^E]|$)", ""] # remove season tag without episode (search doesn't support it)
      - name: diacritics
        args: replace
      # # most ITA TV torrents are in XXxYY format, so we search without S/E prefixes and filter later
      # - name: re_replace
      #   args: ["S0?(\\d{1,2})", " $1 "]
      # - name: re_replace
      #   args: ["E(\\d{2,3})", " $1 "]
    inputs:
      search: "{{ .Keywords }}"
      category: "{{range .Categories}}{{.}};{{end}}"
      page: torrents
      active: 0
    rows:
      selector: div.b-content > table > tbody > tr > td > table.lista > tbody > tr:has(a[href^="index.php?page=torrents&category="])
    fields:
      title:
        selector: td:nth-child(2) > a
        # normalize to SXXEYY format
        # filters:
        #   - name: re_replace # replace special characters with " " (space)
        #     args: ["[^a-zA-Z0-9]|\\.", " "]
        #   # normalize to SXXEYY format
        #   - name: re_replace
        #     args: ["(\\d{2})x(\\d{2})", "S$1E$2"]
        #   - name: re_replace
        #     args: ["(\\d{1})x(\\d{2})", "S0$1E$2"]
        #   - name: re_replace #Stagione X --> S0X
        #     args: ["Stagione (\\d{0,1}\\s)", "S0$1"]
        #   - name: re_replace #Stagione XX --> SXX
        #     args: ["Stagione (\\d{2}\\s)", "S$1"]
        #   - name: re_replace #/ Episodio [YY-YY --> EYY-YY
        #     args: ["(\\s\\/\\sEpisodio|\\s\\/\\sEpisodi|\\sEpisodio|\\s\\|\\sEpisodio|\\sEpisodi)\\s\\[", "E"]
        #   - name: re_replace #/ Completa [episodi  YY-YY --> EYY-YY
        #     args: ["(\\s\\/\\sCompleta\\s\\[episodi\\s)", "E"]
        #   - name: re_replace #remove  di YY] | remove /YY]
        #     args: ["(\\sdi\\s\\d{1,2}|\\/\\d{1,2})\\]", " "]
        #   - name: re_replace #remove various
        #     args: ["(Serie completa|Completa|\\[in pausa\\])", ""]
          # fine prova
      download: # handle torrents with normal torrent file download
...
```
So why you think is normal? At this point I think I've misunderstood some behavior about the code.
My expectation is
I search Grimm S06 -> the script remove S0 with the keywordfilter -> the real search became Grimm 6

2. About the second discussion, do you think is better to make one generic search and then try to adjust the results with the filters? So for example if I need to search:
Grimm S04 -> The search became Grimm 4
Grimm S04E11 -> The search became Grimm 4 11
Then I've to apply filter to all results, I understood well?

My original idea/question is to make 2 separate search with different keywordfilter so:
on first search I've Grimm S04 -> and during a search became Grimm S04
on first search I've Grimm S04E11 -> and during a search became Grimm S04E11
because of keyfilter and
on second search I've Grimm S04 -> and during a search became Grimm 4
on second search I've Grimm S04E11 -> and during a search became Grimm 4x11
always because keyfilter
So, basically I've 2 search with different keyfilters in order to make 2 different search, after that I've to filter the results to transform the data into S01E01 or S01

But you told me is not possible make 2 search into a single definition file, right?

So the very fist problem to solve/understand is, why if I set a keywordsfilters that transmofr S04 into 4, why I can see that Jackett search S04? On the log I linked to you there's also a piece of code where you cann seee that the research was S04 and not 4

Any idea?

**Comment by kaso17** [comment]
I found your problem.
You have to use a upper case S in your manual search ;)

Il Corsaro Blu already has some similar here:
https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Definitions/ilcorsaroblu.yml#L61-L64

I was wrong. It should be possible to make this work with two searches.
You'll have to use multiple paths (same page) and override the search input accordingly.
See this example of isohunt which uses this to get the first the pages of the results:
https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Definitions/isohunt2.yml#L53-L62

Within the override you can use only a single regex
`search: "{{ re_replace .Keywords "..." "..." }}"`
But for doing a season/episode transformation this should be possible with one single (more complex) regex.

**Comment by Jorman** [comment]
I'll try with your isohunt example, thanks for the info.

About the lower/upper case, is not possible make every query case
insensitive? I searched a bit but I haven't found a solution that works in
this case. Or the other options is to make all upper or lower and then make
all filters upper or lower

What do you think?

Il ven 3 ago 2018, 18:08 kaso17 <notifications@github.com> ha scritto:

**Comment by Jorman** [comment]
Ok, I adjusted the regex and you right, case sensitive. At this point I've some question for you
1. Is possible make all case insensitive by default?
2. I think the 2 search way is not the best, to many possible search to handle. In other hand if I only search arrow 6 instead arrow S06 I find a lot of unwanted results, like arrow s02e06 and so on... Maybe do you have a better idea?
3. How much important is the categories into Jackett? I mean, on Il corsaro Nero I see that the latest torrent are grabbed, but what happen during an automatic search from sonarr? Or during the rss sync? The categories are requested? Because I see I can make more precise results if I search inside the right  categories.

**Comment by kaso17** [comment]
@Jorman 
1) you can make regexes case insensitive by adding `(?i)` at the beginning.
e.g. `args: ["(?i)\\s(mkv|avi|divx|xvid|mp4)$", ""]`
2) The andmatch filter should take care of most of your problems. It's a special filter on the row level (has no effect in other contexts. It's not used correctly in your most recent changes.
See e.g. here for an example:
https://github.com/Jackett/Jackett/blob/a3c685388eae9389d1f29be7671250d5ca957b95/src/Jackett.Common/Definitions/eztv.yml#L28-L29
3) assuming the indexer supports searching in categories (not all do) this will improve your search results as you already mentioned. Sonarr specifies 5030 (TV/SD) and 5040 (TV/HD) by default but you can adjust it accordingly.

**Comment by Jorman** [comment]
Thanks for the case insensitive, I'll change all script to make insensitive, make no sense for me do case sensitive, I don't know any tracker that make sensitive difference.

So I've to apply the andmatch filter right under the row selector?

About the search, ok some tracker support it, for example il corsaro blu and shareisland and others, but I don't know how to tell to the script that I want to search on category if a category was selected.

**Comment by kaso17** [comment]
I see that you already fixed the andmatch filter usage.

Regarding the categories: You just have to translate the categories by somehow using the .Categories variable (usually via a range expression) to populate the corresponding arguments.

**Comment by Jorman** [comment]
Thank you for your support! For now seems to works
